tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -40,236 +87,301 @@ auto read_json(std::filesystem::path const& path) {
-    ComplexValue<sym> const result = mag * exp(1.0i * angle);
-    ComplexValue<sym> const result_ref = mag_ref * exp(1.0i * angle_ref);
-    if constexpr (is_symmetric_v<sym>) {
-        return cabs(result - result_ref) < (cabs(result_ref) * rtol + atol);
+    if constexpr (std::is_same_v<std::decay_t<T>, double>) {
+        std::complex<double> result{magnitude * exp(1.0i * angle)};
+        std::complex<double> ref_result{ref_magnitude * exp(1.0i * ref_angle)};
+        return std::abs(result - ref_result) < (std::abs(ref_result) * rtol + atol);
     } else {
-        return (cabs(result - result_ref) < (cabs(result_ref) * rtol + atol)).all();
+        return ref_angle == angle && ref_magnitude == magnitude;
     }
 }
-
-bool check_angle_and_magnitude(RawDataConstPtr reference_result_ptr, RawDataConstPtr result_ptr,
-                               MetaAttribute const& angle_attr, MetaAttribute const& mag_attr, double atol, double rtol,
-                               Idx obj) {
-    if (angle_attr.ctype == CType::c_double) {
-        assert(mag_attr.ctype == CType::c_double);
-        return check_angle_and_magnitude<symmetric_t>(reference_result_ptr, result_ptr, angle_attr, mag_attr, atol,
-                                                      rtol, obj);
+// can probably combine with check_angle_and_magnitude above
+template <typename T>
+bool compare_value(T const& ref_attribute_value, T const& attribute_value, double atol, double rtol) {
+    if constexpr (std::is_same_v<std::decay_t<T>, std::array<double, 3>>) {
+        // std::ranges::all_of make it better, it is not readable
+        return std::transform_reduce(ref_attribute_value.begin(), ref_attribute_value.end(), attribute_value.begin(),
+                                     Idx{}, std::plus<Idx>{},
+                                     [&](double element, double ref_element) {
+                                         return std::abs(element - ref_element) < std::abs(ref_element) * rtol + atol
+                                                    ? Idx{1}
+                                                    : Idx{};
+                                     }) == Idx{3}
+                   ? true
+                   : false;
+    } else if constexpr (std::is_same_v<std::decay_t<T>, double>) {
+        return std::abs(attribute_value - ref_attribute_value) < (std::abs(ref_attribute_value) * rtol + atol);
+    } else {
+        return ref_attribute_value == attribute_value;
     }
-    assert(angle_attr.ctype == CType::c_double3);
-    assert(mag_attr.ctype == CType::c_double3);
-    return check_angle_and_magnitude<asymmetric_t>(reference_result_ptr, result_ptr, angle_attr, mag_attr, atol, rtol,
-                                                   obj);
 }
 
 // assert single result
-void assert_result(ConstDataset const& result, ConstDataset const& reference_result,
+void assert_result(OwningDataset const& owning_result, OwningDataset const& owning_reference_result,
```

Comment:
I personally wouldn't, unfortunately. Even though it is humongoes, there are just many things that must be done in order to check the results, and at least this way is very linear, because it will be very complex nevertheless.

On the other hand, I could just make the lambdas into proper functions, which would reduce the length of this one by a lot and not necessarily increase complexity, but that comes at the cost of having huge and ugly signatures for the resulting functions. Is that something we want to do?

Reply:
Cleaned it up